new/cookbook-segment-deprecation.md

```diff
@@ -0,0 +1,51 @@
+---
+RFC: unassigned
+Author: Lamont Granquist <lamont@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Title
+
+Cookbook Segment Deprecation
+
+## Motivation
+
+    As a Chef User/Developer,
+    I want to be able to extend the structure of cookbooks without fighting with a wire protocol
+    written 6 years ago,
+    Because nobody wants that.
+
+## Specification
+
+This RFC will add a special `all_files` segment to cookbooks which will duplicate (initially)
+all of the existing cookbook manifest along with arbitrary files outside of the current scope of
+any of the cookbook segments.  The Chef Server will need to be patched in order to support this
+new 'segment' and accept it since the Chef Server currently does strict enforcement of file
+segments server-side.  This will also generate a version bump of the Chef Server protocol which
+will be used by clients to determine if the Chef Server supports this field or not.
+
+New knife clients that support this segment will use the Chef Server protocol version to determine
+if they MUST upload cookbooks using the old segment format, or if they SHOULD upload using only
+the `all_files` format.
+
+For backwards compatibility, the Chef Server MUST serve cookbooks with both standard segments and the
+new `all_files` segment unless it can determine that the client prefers one or other.  If it can
+be determined then the server MUST serve cookbooks to old clients with the old format, and SHOULD
+serve cookbooks to new clients with only the `all_files` segment.
+
+New Chef Clients MUST accept either old style segments, cookbooks with the new `all_files` segment
+or cookbook metadata with both styles (and SHOULD favor the `all_files` segment).
+
+This will achieve backwards compatibility and the ability for knife, chef-client and the chef-server
```

Comment:
I'm hoping that the implementation is at a deep enough layer in the objects that client code won't have to be updated everywhere.